### PR TITLE
Eager migration of workflow's database

### PR DESCRIPTION
**Pull request by gdraynz**

### CODE REVIEW

File: nyuki/workflow/db/migrations/0001_structure_change.py [hunk 1]
```diff
@@ -103,20 +107,21 @@ def __init__(self, host, database, **kwargs):
...
-        log.info('Workflow templates splited and migrated')
+        log.info('%s workflow templates splited and migrated', count)
 
     @timed
     async def _migrate_workflow_instances(self):
         """
         Replace workflow instance documents with new structure.
         """
-        i = 0
+        i, count = (0,) * 2
```
Comment (Ketouem) [inline nyuki/workflow/db/migrations/0001_structure_change.py]: `i` seems to be useless here

File: nyuki/workflow/db/migrations/0001_structure_change.py [hunk 2]
```diff
@@ -112,23 +112,33 @@ def __init__(self, host, database, **kwargs):
...
             instance = workflow.pop('exec')
             instance['_id'] = workflow.pop('_id')
             instance['template'] = workflow
-            bulk.find({'_id': instance['_id']}).upsert().replace_one(instance)
+            bulk.insert(instance)
+            old_bulk.find({'_id': instance['_id']}).remove_one()
 
             if i == 1000:
-                await bulk.execute()
+                await asyncio.wait([
```
Comment (Ketouem) [inline nyuki/workflow/db/migrations/0001_structure_change.py]: Completely does the job (:+1: for thinking about the incomplete batch cleaning with `if i > 0`) but it seems there is a lot of duplicated code scattered across this migration. Can the process of looping over the old collection to populate the new one be factorised into an inner method/function ?

File: nyuki/workflow/db/migrations/0001_structure_change.py [hunk 3]
```diff
@@ -220,12 +221,13 @@ def _new_task(self, task, workflow_instance_id=None):
...
             bulk.find({'_id': instance['_id']}).upsert().replace_one(instance)
 
-            task_bulk = tasks_col.initialize_unordered_bulk_op()
-            for task in tasks:
-                task = self._new_task(task, instance['_id'])
-                # This is an older collection, so we don't use '_id'.
-                task_bulk.find({'id': task['id']}).upsert().replace_one(task)
-            await task_bulk.execute()
+            try:
+                await tasks_col.insert_many([
```
Comment (Ketouem) [inline nyuki/workflow/db/migrations/0001_structure_change.py]: 👍  do you have a rough estimation about the gain ?

File: nyuki/workflow/db/storage.py [hunk 4]
```diff
@@ -47,6 +47,16 @@ def configure(self, host, database, **kwargs):
         self.lookups = DataProcessingCollection(self._db, 'lookups')
         self.triggers = TriggerCollection(self._db)
 
+    async def index(self):
+        await self._workflow_templates.index()
```
Comment (Ketouem) [inline nyuki/workflow/db/storage.py]: 👍  

> [Explicit is better than implicit](https://www.python.org/dev/peps/pep-0020/#id3)

File: nyuki/workflow/db/migrations/0001_structure_change.py (hunk 3 above)
Comment (gdraynz) [reply nyuki/workflow/db/migrations/0001_structure_change.py]: `insert_many()` is basically a bulk with multiple `task_bulk.insert(task)` inside.

File: nyuki/workflow/db/migrations/0001_structure_change.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+    def _migrate_one_task_template(self, template):
+        """
+        Do the task configuration migrations.
+        Add the new 'timeout' field.
+        """
+        config = template['config']
+        template['timeout'] = None
+
+        if template['name'] == 'join':
+            template['timeout'] = config.pop('timeout', None)
```
Comment (lhavel) [inline nyuki/workflow/db/migrations/0001_structure_change.py]: :+1:

File: nyuki/workflow/db/storage.py [hunk 6]
```diff
@@ -47,6 +47,16 @@ def configure(self, host, database, **kwargs):
...
+        await self._task_templates.index()
+        await self._metadata.index()
```
Comment (lhavel) [inline nyuki/workflow/db/storage.py]: Align your code with the actual name of the collections: `metadata` => `workflow_metadata`

File: nyuki/workflow/db/migrations/__init__.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import os
+from importlib import util, machinery
```
Comment (lhavel) [inline nyuki/workflow/db/migrations/__init__.py]: unused `machinery` import

File: nyuki/workflow/db/migrations/0001_structure_change.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    def __init__(self, host, database, **kwargs):
+        client = AsyncIOMotorClient(host, **kwargs)
+        self.db = client[database]
+
+    @timed
+    async def run(self):
+        """
+        Run the migrations.
+        """
+        log.info('Starting migrations')
```
[the remark this answers is not in the document]
Comment (lhavel) [reply nyuki/workflow/db/migrations/0001_structure_change.py]: The same applies to the _end-of-migration_ log message.

**Pull request merged**